### PR TITLE
kubernetes-kcp: init at 0.26.0

### LINKED ISSUE #357434: Package request: kcp
## Project description

kcp is a Kubernetes-like control plane focusing on:

- A control plane for many independent, isolated “clusters” known as workspaces
- Enabling API service providers to offer APIs centrally using multi-tenant operators
- Easy API consumption for users in their workspaces

kcp can be a building block for SaaS service providers who need a massively multi-tenant platform to offer services to a large number of fully isolated tenants using Kubernetes-native APIs. The goal is to be useful to cloud providers as well as enterprise IT departments offering APIs within their company.

## Metadata

* homepage URL: https://www.kcp.io
* source URL: https://github.com/kcp-dev/kcp
* license: Apache License 2.0
* platforms: linux, darwin

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by rytswd**
Introduces [kcp](https://kcp.io), Kubernetes-like control planes for form-factors and use-cases beyond Kubernetes and container workloads.

Closes #357434 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (steeleduncan) [inline pkgs/applications/networking/cluster/kcp/default.nix:63]: `with lib` is deprecated #208242

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (steeleduncan) [inline pkgs/applications/networking/cluster/kcp/default.nix:20]: `refs/tags/v${version}`

**Review by steeleduncan — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (rytswd) [reply pkgs/applications/networking/cluster/kcp/default.nix:63]: Ah thanks for the catch, I checked out other Kubernetes related package and got it over here. This is fixed in the latest commit (squashed / force-pushed).

**Comment by rytswd** [comment]
I'm not sure what the next step should be -- do I need to tag someone who has write access to merge the PR?

**Comment by steeleduncan** [comment]
Yes, someone with the commit bit will pass by eventually. You could look for someone who often merges go modules and request a review

By the way, it probably makes sense to
- remove the various TODOs and comments in the derivation
- specify platforms in the meta? (I don't know if this is common for go derivations)

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (Aleksanaa) [inline pkgs/applications/networking/cluster/kcp/default.nix:1]: This should be in `pkgs/by-name/ku/kubernetes-kcp`

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (Aleksanaa) [inline pkgs/applications/networking/cluster/kcp/default.nix:21]:
```suggestion
    hash = "sha256-ZEgDeILo2weSAZgBsfR2EQyzym/I/+3P99b47E5Tfrw=";
```

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (Aleksanaa) [inline pkgs/applications/networking/cluster/kcp/default.nix:36]: No need to, just mark unkown

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (Aleksanaa) [inline pkgs/applications/networking/cluster/kcp/default.nix:48]:
```suggestion
  postInstall = lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform)''
```

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (Aleksanaa) [inline pkgs/applications/networking/cluster/kcp/default.nix:65]:
```suggestion
    description = "Kubernetes-like control planes for form-factors and use-cases beyond Kubernetes and container workloads";
```

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (rytswd) [reply pkgs/applications/networking/cluster/kcp/default.nix:36]: @Aleksanaa Could you elaborate why that's the case? Is this just customary, and/or because it's easier to update the reference for updates by simply using the tag?
This will be a slight diversion from the original binary, and when it is as simple as providing a commit reference, I thought it's worth keeping it around. (Also we should technically have the Kubernetes major / minor version reference, which would ensure the Nix package will be the same as the official binary.)

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (rytswd) [reply pkgs/applications/networking/cluster/kcp/default.nix:1]: Thanks, I wasn't too clear about which one to use, so placed it near other Kubernetes related packages such as Helm and k3s. I can see many other packages under `pkgs/by-name/ku/`, which looks like a better place to store kcp setup 🙏

**Comment by rytswd** [comment]
Thanks for the pointer, @steeleduncan -- regarding your points:

> By the way, it probably makes sense to
> 
> - remove the various TODOs and comments in the derivation
> - specify platforms in the meta? (I don't know if this is common for go derivations)

I kept the TODO items as those are something we should handle for proper feature parity (it may be a stretch to call this a feature, but nonetheless a difference from the official binary).
As to the platform, this should work on both nixos and darwin, so I suppose this can be left as is.

**Comment by rytswd** [comment]
@Aleksanaa Thanks for the review! I made the suggested changes and a few adjustments (namely argument for `stdenv` and additional binaries of kubectl plugins).

I also went ahead to squash the code change commits into a single commit. My understanding is that I will need to keep the contributor addition commit separate, so that was kept as is. Rebase / squash / force-push made it difficult to see the diff, but I hope this one is simple enough to wrap your head around the change 🙏

File: pkgs/applications/networking/cluster/kcp/default.nix
Comment (Aleksanaa) [reply pkgs/applications/networking/cluster/kcp/default.nix]: Because setting rev specifically will make automatic update wrong, and not very useful as the version is sufficient to find the revision

Comment (rytswd) [reply pkgs/applications/networking/cluster/kcp/default.nix]: Thanks for the clarification! The automatic update support would be nice to have, and indeed the tag should give us the revision. But it is also possible that the tag gets removed and recreated for a different commit (though unlikely), and this revision reference acts as a safety net in a way I suppose.

In either case, it is also not a fast moving project AFAIAA, and it is something I can revisit later. The change has been made to remove the unnecessary let binding and git commit reference kept as unknown.

**Comment by rytswd** [comment]
@Aleksanaa Sorry to bother you multiple times, I think I have worked out most parts covered in the previous review. Would you be able to review and merge this when you have a moment? There are a few TODOs that I purposely left, which we can tackle later 🙏

File: pkgs/by-name/ku/kubernetes-kcp/package.nix
Comment (pluiedev) [inline pkgs/by-name/ku/kubernetes-kcp/package.nix:31]: You could use `lib.versions.major kubernetes.version` probably (same goes for minor)

Comment (rytswd) [reply pkgs/by-name/ku/kubernetes-kcp/package.nix:31]: Ah thanks for the note, I'll check more into that! It is probably going to be a bit more involved, though, as this needs to pick up the version details from `go.mod`. That probably has to be done preInstall phase, or even by adjusting the installPhase. I do not think this is a must have, and we could look to improve and match the upstream deployment setup later down the line

**Comment by rytswd** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 357444`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>kubernetes-kcp</li>
  </ul>
</details>

**Review by Aleksanaa — APPROVED** [review]

**Pull request merged**
